widgets/src/qsharp_widgets/__init__.py

```diff
@@ -35,3 +36,46 @@ class EstimateDetails(anywidget.AnyWidget):
     def __init__(self, estimates):
         super().__init__()
         self.estimates = estimates
+
+
+class Histogram(anywidget.AnyWidget):
+    _esm = pathlib.Path(__file__).parent / "static" / "index.js"
+    _css = pathlib.Path(__file__).parent / "static" / "index.css"
+
+    comp = traitlets.Unicode("Histogram").tag(sync=True)
+    buckets = traitlets.Dict().tag(sync=True)
+    shot_count = traitlets.Integer().tag(sync=True)
+
+    _new_buckets = {}
+    _new_count = 0
+    _last_message = time.time()
+
+    results = []
+
+    def _update_ui(self):
+        self.buckets = self._new_buckets.copy()
+        self.shot_count = self._new_count
+        self._last_message = time.time()
+
+    def add_result(self, result):
+        result_str = str(result["result"])
+        old_value = self._new_buckets.get(result_str, 0)
+        self._new_buckets.update({result_str: old_value + 1})
+        self._new_count += 1
+
+        # Only update the UI max 20 times per second
+        if time.time() - self._last_message >= 0.05:
+            self._update_ui()
+
+    def __init__(self, entry_expr=None, shot_count=None):
+        super().__init__()
+        if entry_expr and shot_count:
+            self.run(entry_expr, shot_count)
+
+    def run(self, expr, shots):
+        import qsharp
```

Comment:
Now that this imports qsharp, should qsharp be listed as a dependency of the package? Alternatively, this could do a try-catch and wrap the exception with some additional help text to tell the user that histograms depend on qsharp and they need to ensure it is installed properly.

Reply:
Adding as a dependency might make sense. I really don't see how you'd use qsharp_widgets without qsharp. I expect most folks will get both at once anyway via the optional extra on the qsharp package with `pip install "qsharp-lang[widgets]"`

Reply:
I agree, though now that you mention the optional dependency I'm curious how pip handles what looks like a cycle there... `pip install 'qsharp-lang[widgets]'` will install qsharp-lang and then qsharp-widgets, but qsharp-widgets depends on qsharp-lang...

Reply:
Related: I think we should do the rename from `qsharp-lang` to `qsharp` this week, so I think it's okay to go ahead and put the new name, `qsharp`, as a dependency now.

Reply:
If we get multi-file in, then I wouldn't mind pushing out another dev build. Might we weird if this preview QDK starts publishing builds to the old 'qsharp' pip package now.